File: src/main/java/com/morak/performancetracker/aop/web/PerformanceInterceptor.java
package com.morak.performancetracker.aop.web;

import com.morak.performancetracker.context.Accumulator;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import org.springframework.stereotype.Component;
import org.springframework.web.servlet.HandlerInterceptor;
import org.springframework.web.servlet.ModelAndView;

@Component
public class PerformanceInterceptor implements HandlerInterceptor {

    private final Accumulator accumulator;
    private final WebMonitor webMonitor;

    public PerformanceInterceptor(Accumulator accumulator, WebMonitor webMonitor) {
        this.accumulator = accumulator;
        this.webMonitor = webMonitor;
    }

    @Override
    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) {
        if (!isPreflight(request)) {
            webMonitor.start(request);
        }
        return true;
    }

    @Override
    public void postHandle(HttpServletRequest request, HttpServletResponse response, Object handler,
                           ModelAndView modelAndView) {
        if (isPreflight(request) || webMonitor.getUri() == null) {
            return;
        }
        webMonitor.end();
        accumulator.add(webMonitor);
    }

    private boolean isPreflight(HttpServletRequest request) {
        return request.getMethod().equals("OPTIONS");
    }
}

File: src/main/kotlin/com/morak/performancetracker/aop/web/PerformanceInterceptor.kt
package com.morak.performancetracker.aop.web

import com.morak.performancetracker.context.Accumulator
import org.springframework.http.HttpMethod
import org.springframework.lang.Nullable
import org.springframework.stereotype.Component
import org.springframework.web.servlet.HandlerInterceptor
import org.springframework.web.servlet.ModelAndView
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@Component
class PerformanceInterceptor(private val accumulator: Accumulator, private val webMonitor: WebMonitor) :
    HandlerInterceptor {
    override fun preHandle(request: HttpServletRequest, response: HttpServletResponse, handler: Any): Boolean {
        if (!isPreflight(request)) {
            webMonitor.start(request)
        }
        return true
    }

    override fun postHandle(
        request: HttpServletRequest,
        response: HttpServletResponse,
        handler: Any,
        @Nullable modelAndView: ModelAndView?
    ) {
        if (isPreflight(request) || webMonitor.uri == null) {
            return
        }
        webMonitor.end()
        accumulator.add(webMonitor)
    }

    private fun isPreflight(request: HttpServletRequest): Boolean {
        return request.method == HttpMethod.OPTIONS.name
    }
}
